### PR TITLE
feat: semver compatibility check

### PR DESCRIPTION
**Pull request by floredelasteyrie**

### CODE REVIEW

File: src/outcome/utils/semver.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    a_parsed = semver.VersionInfo.parse(a)
+    b_parsed = semver.VersionInfo.parse(b)
+
+    if a_parsed.major != b_parsed.major:
+        return Compatibility.incompatible
+
+    elif a_parsed.minor != b_parsed.minor or a_parsed.patch != b_parsed.patch:
+        return Compatibility.compatible
+
+    return Compatibility.equal
```
Comment (d-tw) [inline src/outcome/utils/semver.py:22]: Do we actually need `equal`? I'm not sure that's a state of compatibility

We'd end up with code like 

```py
compat = compatible(a,b)

if a !== Compatibility.incompatible:
    ...
```

File: src/outcome/utils/semver.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+def compatible(a: str, b: str) -> Tuple[bool, Union[None, str]]:
+    a_parsed = semver.VersionInfo.parse(a)
+    b_parsed = semver.VersionInfo.parse(b)
+
+    if a_parsed.major != b_parsed.major:
+        return Compatibility.incompatible
+
+    elif a_parsed.minor != b_parsed.minor or a_parsed.patch != b_parsed.patch:
+        return Compatibility.compatible
```
Comment (d-tw) [inline src/outcome/utils/semver.py:20]: How about a unit test? :)

**Review by d-tw — CHANGES_REQUESTED** [review]

File: src/outcome/utils/semver.py (hunk 1 above)
Comment (floredelasteyrie) [reply src/outcome/utils/semver.py:22]: I need to know if they're `compatible` and not `equal`, but I don't actually need the `equal` case.
What would I return in that case ?

Comment (d-tw) [reply src/outcome/utils/semver.py:22]: I consider it to be two separate pieces of logic.

I would change your `utils` code to be:

```py

class Compatibility(Enum):
    compatible = 'compatible'
    incompatible = 'incompatible'

    # A third case, for potential future use 
    # if we decide to check build or pre-release info
    unknown = 'unknown'


def compatibility(a: str, b: str) -> Compatibility:
    a_parsed = semver.VersionInfo.parse(a)
    b_parsed = semver.VersionInfo.parse(b)

    if a_parsed.major != b_parsed.major:
        return Compatibility.incompatible

    return Compatibility.compatible


def are_compatible(a: str, b: str) -> bool:
    return compatibility(a, b) == Compatibility.compatible
```

Then in your other code, have the following:

```py
if a != b and are_compatible(a, b):
    ...
```

**Review by d-tw — APPROVED** [review]

**Review by d-tw — APPROVED** [review]

**Pull request merged**